### PR TITLE
Feat/testing-basics-garden-project

### PR DESCRIPTION
**Pull request by mesporas1**
This PR contains the code for the Basics in Testing project to address #100.

@scissorsneedfoodtoo I started working on the steps for the Basics in Testing Project. I wanted to see if my initial steps are on the right track before I write the rest of the steps. Appreciate your feedback!

### CODE REVIEW

**Comment by mesporas1** [comment]
@scissorsneedfoodtoo I started working on the steps for the Basics in Testing Project. I wanted to see if my initial steps are on the right track before I write the rest of the steps. Appreciate your feedback!

**Comment by scissorsneedfoodtoo** [comment]
@mesporas1, awesome! I skimmed through and everything looks good at first glance. Let's merge this for now and I'll go through this again and leave more detailed comments ASAP.

File: testing-basics-garden-project/garden-test001.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+/* We will be writing tests against the code for the Garden program found in the starter-files directory.
+   Let's write tests against the Garden requirements first.
+   To start, import the Garden object from the src directory
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test001.js:4]: This reads really well. Short, sweet, and to the point.

It's probably review at this point, but on the off chance that someone skips previous projects, we can include a quick example of importing an object or module. Also made a couple of minor formatting tweaks:

```
/* 
  We will be writing tests against the code for the `Garden` program found in the starter-files directory.
  Let's write tests against the `Garden` requirements first.
  To start, import the `Garden` object from the `src` directory.
  For example, here's how you would import `Plane` from the same directory:
  `import { Plane } from "../src/Plane";`
*/
```

File: testing-basics-garden-project/garden-test002.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import { Garden } from '../src/Garden';
+
+/* Often, code is tested against assert libraries.
+   Javscript has its own assert funcitons that you can look into.
+   For this project, we're going to use the chai assert library.
+   Require the "chai" module in your code and set it to variable "chai"
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test002.js:7]: Again, you have a knack for writing these test descriptions. Hope you don't mind, but I tried to shorten it slightly to fit another quick example in, just on the off chance that someone skipped ahead to this project:

```
/* 
  Often, code is tested against assert libraries.
  JavaScript has its own assert functions that you can look into, but for this project, we're going to use the Chai Assertion Library.
  Require the `"chai"` module in your code and set it to variable `chai`.
  For example, here's how to require Node's File System module:
  `const fs = require("fs");`
*/
```

File: testing-basics-garden-project/garden-test003.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { Garden } from '../src/Garden';
+const chai = require("chai");
+
+/* We are going to use the command "assert" a lot throughout this project.
+   To use chai's assert, you'd have to call chai.assert.
+   Typing that for every test would be tedious, so we won't do that.
+   Instead, create a variable called "assert" and set it to chai's assert command.
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test003.js:8]:
```
/* 
  We are going to use the command `assert` a lot throughout this project.
  To use Chai's `assert`, you'd have to call `chai.assert`.
  Typing that for every test would be tedious, so we won't do that.
  Instead, create a variable called `assert` and set it to Chai's assert command.
*/
```

File: testing-basics-garden-project/garden-test004.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import { Garden } from '../src/Garden';
+const chai = require("chai");
+const assert = chai.assert;
+
+// Define a new Garden object called myGarden using let instead of var or const
+// I'll explain why later..
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test004.js:6]: We could probably just shorten this to the first sentence:

```
Define a new `Garden` object called `myGarden` using `let` instead of `var` or `const`.
```

File: testing-basics-garden-project/garden-test005.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+const chai = require("chai");
+const assert = chai.assert;
+
+let myGarden = new Garden();
+
+/* Let's use our assert function by testing if a Garden object is an instance of Garden.
+   For example, if I create a "Plane" object with the variable "myPlane".
+   I can assert that "myPlane" is a "Plane" object with "assert.instanceOf(myPlane, Plane)"
+   Write an assert statement checking if the variable myGarden is an instance of Garden
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test005.js:11]: This step is great. Really clear example and instructions. I just made some minor tweaks so everything will be rendered as code on our platform later.

```
/* 
  Let's use our assert function by testing if a `Garden` object is an instance of `Garden`.
  For example, if I create a `Plane` object with the variable `myPlane`,
  I can assert that `myPlane` is a `Plane` object with `assert.instanceOf(myPlane, Plane)`.
  Write an `assert` statement checking if the variable `myGarden` is an instance of `Garden`.
*/ 
```

File: testing-basics-garden-project/garden-test006.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+/* Now that we can test our code with assert
+   Let's organize our code into test cases
+   For example, if I want to describe that "myPlane" should be a "Plane",
+   I can write:
+      it("should be a plane", function(){
+         assert.instanceOf(myPlane, Plane)
+      })
+   Write a test case for the garden using "it" and use "should be able to create a Garden"
+   as the string describing the test
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test006.js:18]:
````
/* 
  Now that we can test our code with `assert`, let's organize our code into test cases
  For example, if I want to describe that `myPlane` should be a `Plane`,
  I can write:
  ```
      it("should be a plane", function(){
         assert.instanceOf(myPlane, Plane)
      })
  ```
  Write a test case for the garden using `it` and use `"should be able to create a Garden"` as the string describing the test.
*/
````

File: testing-basics-garden-project/garden-test007.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+   With our "plane" example, if I wanted to group the tests into "Plane Tests",
+   I can write this:
+      describe("Plane tests", function(){
+         it("should be a plane", function(){
+         assert.instanceOf(myPlane, Plane)
+         })
+      })
+   Put the test that you currently have into a test block using "describe"
+   Call the test block "Garden tests"
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test007.js:22]:
````
/* 
  Now we can describe our tests and our test framework will describe which tests are passing.
  We can organize our tests further by grouping them with `describe`.
  With our `Plane` example, if I wanted to group the tests into `"Plane Tests"`, I can write this:
  ```
      describe("Plane tests", function(){
         it("should be a plane", function(){
           assert.instanceOf(myPlane, Plane)
         })
      })
  ```
   Put the test that you currently have into a test block using `describe`.
   Call the test block `"Garden tests"`.
*/
````

File: testing-basics-garden-project/garden-test008.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+      assert.instanceOf(myGarden, Garden);
+   });
+});
+
+/* Great! Now we can group tests into test blocks with "describe"
+   Let's test against more requirements.
+   A garden should be able to add different kinds of plants. Let's start with Fruit Trees!
+   Import "FruitTree" from the "FruitTree.js" file
+   Declare a variable myFruitTree and assign it to a new FruitTree with "Strawberry" as a parameter inside of the "describe" block
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test008.js:17]: Again, great writing here. However we may need to split this up into two steps.

The first could read:

```
/* 
  Great! Now we can group tests into test blocks with `describe`.
  Let's test against more requirements.
  A garden should be able to add different kinds of plants. Let's start with Fruit Trees!
  Import `FruitTree` from the `FruitTree.js` file in `src`.
*/
```

And the next step, maybe called `garden-test008a.js`, could read:

```
// Inside the `describe` block, declare a variable `myFruitTree` and assign it to a new `FruitTree` with `"Strawberry"` as a parameter.
```

File: testing-basics-garden-project/garden-test009.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+   let myGarden = new Garden();
+   let myFruitTree = new FruitTree("Strawberry");
+   it("should be able to create a Garden", function () {
+      assert.instanceOf(myGarden, Garden);
+   });
+});
+
+/* Now, let's create a test case where the Garden should be able to add fruit trees
+   Define the test case with the string "should be able to add fruit trees" and pass in an empty function
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test009.js:16]:
```
/* 
  Now, let's create a test case where the `Garden` should be able to add fruit trees.
  Define the test case with the string `"should be able to add fruit trees"` and pass in an empty function.
*/
```

File: testing-basics-garden-project/garden-test010.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+   });
+
+   it("should be able to add fruit trees", function () {
+
+   })
+});
+
+/*  Add myFruitTree to myGarden in the "should be able to add fruit trees" test case
+    NOTE: Check out how FruitTrees are added to Gardens in the Garden.js file
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test010.js:20]:
```
/*  
  Add `myFruitTree` to `myGarden` in the `"should be able to add fruit trees"` test case.
  NOTE: Check out how `FruitTrees` are added to `Garden`'s in the `Garden.js` file
*/ 
```

File: testing-basics-garden-project/garden-test011.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+   })
+});
+
+/*  Great! Now that we added myFruitTree to myGarden.
+   We can do this using the "isAbove" assert method from Chai.
+   The method works like this:
+      assert.isAbove(a,b) // Checks if a > b
+   Write an assert using "isAbove" to check if the FruitTree was added to the Garden.
+   The assert should check if the length of fruitTrees in "myGarden" is greater than 0.
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test011.js:24]: Changed your example just a bit to make it a little closer to the solution.

```
/*  
  Great! Now that we added `myFruitTree` to `myGarden`, we can write a test to make sure it's working.
  We can do this using the `isAbove` assert method from Chai.
  The method works like this:
  `assert.isAbove(a, 2) // Checks if a > 2`
  Write an assert using `isAbove` to check if the `FruitTree` was added to the `Garden`.
  The assert should check if the length of `fruitTrees` in `myGarden` is greater than 0.
*/
```

File: testing-basics-garden-project/garden-test012.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+      myGarden.addPlant(myFruitTree);
+      assert.isAbove(myGarden.fruitTrees.length, 0);
+   })
+});
+
+/*  Congrats! You have now made two test cases!
+    Let's create another test case called "should not allow more than 5 fruit trees"
+    This will correspond to the next requirement.
+    You don't need to implement the test yet.
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test012.js:23]:
```
/*  
  Congrats! You have now made two test cases!
  Let's create another test case called `"should not allow more than 5 fruit trees"`.
  This will correspond to the next requirement.
  You don't need to implement the test yet.
*/
```

File: testing-basics-garden-project/garden-test013.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+      assert.isAbove(myGarden.fruitTrees.length, 0);
+   })
+
+   it("should not allow more than 5 fruit trees", function () {
+   });
+});
+
+/* Now that we have the test, let's implement it.
+   Write a for loop that adds more than 5 myFruitTree objects to myGarden
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test013.js:24]:
```
/* 
  Now that we have the test, let's implement it.
  Write a for loop that adds more than 5 `myFruitTree` objects to `myGarden`.
*/
```

File: testing-basics-garden-project/garden-test014.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+   })
+
+   it("should not allow more than 5 fruit trees", function () {
+      for (let i = 0; i < 6; i++) {
+         myGarden.addPlant(myFruitTree);
+      }
+   });
+});
+
+// Write an assert statement that checks if the length of fruitTrees in myGarden is at most 5.
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test014.js:25]:
```
// After the `for` loop, use the `isAtMost` method to write an assert statement that checks if the length of `fruitTrees` in `myGarden` is at most 5.
```

File: testing-basics-garden-project/garden-test015.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+   We can setup what happens before each test using "beforeEach"
+   Use "beforeEach" to setup the myGarden variable for every test case
+   e.g.
+      describe("Tests", function(){
+         let initialVar
+         beforeEach(() => {
+            // setup code for initialVar
+         })
+      })
+*/
```
Comment (scissorsneedfoodtoo) [inline testing-basics-garden-project/garden-test015.js:37]: Tried to clarify things a bit and change the example to something that's closer to the final code. Hopefully it's not just giving the answer away. What do you think?

````
/* 
  Our tests pass, but they are not independent of other tests.
  This can cause problems for future tests.
  We can set up what happens before each test using the `beforeEach` hook.
  Just below your `myGarden` variable, use `beforeEach` to set up the `myGarden` variable for every test case, for example:
  ```
  describe("Tests", function(){
    let Plane
    beforeEach(() => {
      Plane = new Plane()
    })
  })
  ```
*/
````

Comment (mesporas1) [reply testing-basics-garden-project/garden-test015.js:37]: Hi Kris, 

This is good! It's more explicit than what I originally have.
I don't think it gives the answer away.

**Comment by mesporas1** [comment]
Thanks for the feedback Kris!
I'll follow the format from your comments for my steps and make a pull request when I add the rest of the steps.
